update to use go1.13.4

liggitt opened the pull request:
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
* Updates to go1.13.4
  * Updates to [bazel rules_go 0.20.2](https://github.com/bazelbuild/rules_go/releases/tag/0.20.2)
  * Updates to [gazelle 0.19.0](https://github.com/bazelbuild/bazel-gazelle/releases/tag/0.19.0)
  * Updates to [buildtools 0.29.0](https://github.com/bazelbuild/buildtools/releases/tag/0.29.0)
* Updates golang.org/x/... dependencies to go1.13 levels
* Updates to github.com/go-openapi/jsonpointer v0.19.3 to pick up https://github.com/go-openapi/jsonpointer/pull/4, required for 1.13
* Updates k8s.io/repo-infra to pick up https://github.com/kubernetes/repo-infra/pull/150, required to make BUILD file generation consistent
* Updates minimum go version to 1.13.4
  * Required for consistent vendor generation, and for bazel to build golang.org/x/crypto
* Continues to build in GOPATH mode (not module mode). This is to minimize build changes so we can pick go1.13 support to release-1.16. Updates to start using modules will be done in follow-ups.

Before merge:
- [x] Wait for fix for https://github.com/golang/go/issues/34978 (1.13.4)
- [x] build/push cross-build image for 1.13.4
- [x] resolve BUILD file issue referencing crypto/ed25519 - https://github.com/bazelbuild/bazel-gazelle/issues/671
- [x] bump golang.org/x/crypto to go1.13 level
- [x] drop TMP commit allowing go1.12.x
- [x] update CI to run with go1.13.4 - https://github.com/kubernetes/test-infra/pull/15113
- [x] Fix flag parsing in node-e2e: https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/82809/pull-kubernetes-node-e2e/1191961293179850752/ and https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/82809/pull-kubernetes-node-e2e-containerd/1191961293142102016/
- [x] update CI to use go1.13.4 images - https://github.com/kubernetes/test-infra/pull/15131
- [x] Fix go version / image used by pull-kubernetes-integration - https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/82809/pull-kubernetes-integration/1191961293163073537/ - https://github.com/kubernetes/test-infra/pull/15144

**Which issue(s) this PR fixes**:
xref https://github.com/kubernetes/kubernetes/issues/82531

**Does this PR introduce a user-facing change?**:
```release-note
Kubernetes now requires go1.13.4+ to build
```

/cc @fejta @cblecker @BenTheElder @thockin

liggitt committed:
opt out of module mode for builds

liggitt committed:
Update build to go1.13.4

cblecker commented:
/sig testing release
/priority important-soon

liggitt committed:
Fix hack/pin-dependency.sh for go1.13

fejta reviewed (approved):
/approve
/lgtm
/hold

BenTheElder commented:
```
2019-09-18 03:00:27.555724 I | integration: terminated 1795862953965652868 (unix://localhost:17958629539656528680)
--- FAIL: TestValidOpenAPISpec (1.42s)
panic: Invalid schema provided to SchemaValidator: object has no key "default" [recovered]
	panic: Invalid schema provided to SchemaValidator: object has no key "default"
```
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/82809/pull-kubernetes-bazel-test/1174153992704364545
hmm
/retest

BenTheElder commented on build/build-image/cross/VERSION:
```diff
@@ -1 +1 @@
-v1.12.9-1
+v1.13.0-1
```
are you planning to push this image?

liggitt replied:
was waiting for the upstream scale issue to be resolved, and 1.13.1 to be released before messing with pushing images

BenTheElder replied:
ACK

zouyee replied:
1.13.1 is out.

cblecker commented:
/lgtm cancel

This PR won't go out as is, as we need go1.13.1

BenTheElder commented:
https://github.com/golang/go/issues/32828#issuecomment-535560654 we still need to wait for 1.13.2

liggitt commented:
go1.13.3 has been released. @wojtek-t is that on the scalability team's radar to verify the perf regression fix on?

wojtek-t commented:
@mm4tt - we should verify the go1.13.3 to check if everything that we needed was cherrypicked there.

mm4tt commented:
I can confirm that golang1.13.3 contains all scalability patches, more details can be found in https://github.com/golang/go/issues/34556 and https://github.com/golang/go/issues/34149.

I think we can go with merging this PR and verifying in our 5K CI/CD tests. We're currently impacted by https://github.com/kubernetes/kubernetes/issues/84151 and the tests are failing but a fix has been already merged and we have high hopes that today's run will pass.

Given that, can we wait till tomorrow with submitting this?

zouyee commented:
At first, `golang version` may be updated in ci-image.
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/82809/pull-kubernetes-typecheck/1187254464956862465

liggitt commented:
@cblecker @BenTheElder sounds like we're ready to do this. what's the process for building/pushing the cross-build image?

BenTheElder commented:
Someone needs to update the image sources for v1.13.3.
A googler needs to do make -C build/build-image/cross push and then file a
promotion CL internally for the resulting image.

On Thu, Oct 24, 2019 at 9:02 AM Jordan Liggitt <notifications@github.com>
wrote:

> @cblecker <https://github.com/cblecker> @BenTheElder
> <https://github.com/BenTheElder> sounds like we're ready to do this.
> what's the process for building/pushing the cross-build image?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/82809?email_source=notifications&email_token=AAHADK4OIIYV43JDYHZIUWDQQHBKVA5CNFSM4IXX6RMKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECFRPGA#issuecomment-545986456>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHADK5HAUUA2CUSDVJOKR3QQHBKVANCNFSM4IXX6RMA>
> .
>

liggitt commented:
/retest

liggitt commented:
seeing a reproducible unit test panic in openapi swagger parsing. digging :(


the following passes on 1.12 and panics on 1.13:
```
import (
  "github.com/go-openapi/validate"
  "github.com/go-openapi/spec"
  "github.com/go-openapi/strfmt"
)

validate.NewSchemaValidator(spec.MustLoadSwagger20Schema(), nil, "", strfmt.Default)
```

liggitt commented:
looks like the reflect.IsZero() added in go1.13 breaks go-openapi/pointer

liggitt committed:
bump github.com/go-openapi/jsonpointer v0.19.3

To fix bug exposed by reflect.Value#IsZero method added in go1.13

liggitt committed:
bump kazel, bazel, gazelle

hack/pin-dependency.sh k8s.io/repo-infra v0.0.1-alpha.1
To fix non-idempotent BUILD file generation

hack/pin-dependency.sh github.com/bazelbuild/bazel-gazelle v0.19.0
hack/pin-dependench.sh github.com/bazelbuild/buildtools 0.29.0
for go1.13.3 support

hack/lint-dependencies.sh

hack/pin-dependency.sh github.com/bazelbuild/buildtools v0.0.0-20190917191645-69366ca98f89
to pick up preferred version for repo-infra

liggitt committed:
bump golang/x/... dependencies

go1.13-levels
hack/pin-dependency.sh golang.org/x/crypto release-branch.go1.13
hack/pin-dependency.sh golang.org/x/net release-branch.go1.13
hack/pin-dependency.sh golang.org/x/tools release-branch.go1.13
hack/pin-dependency.sh golang.org/x/sys release-branch.go1.13

preferred levels for repos without 1.13 branches:
hack/pin-dependency.sh golang.org/x/lint v0.0.0-20190409202823-959b441ac422
hack/pin-dependency.sh golang.org/x/oauth2 v0.0.0-20190604053449-0f29369cfe45
hack/pin-dependency.sh golang.org/x/sync v0.0.0-20190911185100-cd5d95a43a6e
hack/pin-dependency.sh golang.org/x/text v0.3.2
hack/pin-dependency.sh golang.org/x/time v0.0.0-20190308202827-9d24e82272b4

josiahbjorgaard commented:
Bug triage for 1.17 here. Code freeze is Nov. 18th. Will this PR be merged by then? If not, can we move it out of the milestone?

BenTheElder commented:
Ideally this will merge by then.

On Sat, Oct 26, 2019 at 1:41 AM Josiah Bjorgaard <notifications@github.com>
wrote:

> Bug triage for 1.17 here. Code freeze is Nov. 18th. Will this PR be merged
> by then? If not, can we move it out of the milestone?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/82809?email_source=notifications&email_token=AAHADKZMI43DHTOR4XY3RHDQQP7EJA5CNFSM4IXX6RMKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECKDFDY#issuecomment-546583183>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHADK7ETQTCFFJBZQZNYU3QQP7EJANCNFSM4IXX6RMA>
> .
>

BenTheElder commented:
needs rebase

liggitt commented:
rebased

cblecker commented:
/test all
Anything that uses the cross build image or bazel should pass. Things like verify might fail. Let's get some better signal.

liggitt commented:
There's an issue with the BUILD file generated for golang.org/x/crypto referring to a new stdlib package in go1.13 under vendor

liggitt committed:
hack/lint-dependencies.sh

hack/pin-dependency.sh github.com/go-openapi/swag v0.19.5
hack/pin-dependency.sh github.com/gregjones/httpcache v0.0.0-20180305231024-9cad4c3443a7
hack/pin-dependency.sh github.com/mailru/easyjson v0.0.0-20190626092158-b2ccc519800e
hack/pin-dependency.sh github.com/mattn/go-isatty v0.0.9
hack/pin-dependency.sh github.com/pkg/errors v0.8.1
hack/pin-dependency.sh github.com/spf13/pflag v1.0.5
hack/pin-dependency.sh gopkg.in/inf.v0 v0.9.1

liggitt committed:
hack/update-vendor.sh

BenTheElder commented:
/retest

BenTheElder commented:
other than the temp commit at the end:
/lgtm
/approve

liggitt commented:
two green runs of e2e/unit/integration tests at https://prow.k8s.io/pr-history/?org=kubernetes&repo=kubernetes&pr=82809 on e892267 and a3cea6c

dropped temp commit, ensured `hack/verify-typecheck.sh` and `hack/verify-vendor.sh` passed locally

cblecker commented:
/retest

cblecker commented:
/lgtm
/approve
/honk
/woof
/meow

BenTheElder commented:
/woof
/honk
/pony party

BenTheElder commented:
/hold cancel
/lgtm
/approve

cblecker commented:
/hold
Holding to block retesting

liggitt commented:
Added the links to the failed CI runs for those three jobs to the description. I'll look to sig-testing folks to help understand what needs to be adjusted for those jobs..

liggitt committed:
Move e2e_node flag copy into TestMain

As of go1.13, test flags like test.timeout are registered lazily.
This means they are not available in package init() methods:

> Testing flags are now registered in the new Init function,
> which is invoked by the generated main function for the test.
> As a result, testing flags are now only registered when running
> a test binary, and packages that call flag.Parse during package
> initialization may cause tests to fail.

This moves the copy of CLI flags into TestMain, just prior to parse.

BenTheElder commented:
/test pull-kubernetes-integration-podutil

liggitt commented:
/test pull-kubernetes-integration-podutil

BenTheElder commented:
/test pull-kubernetes-integration-podutil

liggitt committed:
Fixup integration tests for TLS1.3

liggitt commented:
/test pull-kubernetes-integration-podutil

liggitt commented:
/test pull-kubernetes-integration-podutil

BenTheElder commented:
/lgtm
/approve
:shipit:

BenTheElder commented:
/test all

BenTheElder commented:
/lgtm

cblecker commented:
/lgtm
/approve
/hold cancel
SHIP IT

liggitt commented:
Current batch merged and this is alone in the merge queue. Restarting the test to test against master HEAD

/test all
